### ISSUE TITLE
Backwards compatibility is broken between 1.9.1 and nightly for `torch.cosine_similarity`

### ISSUE THREAD

**Issue by patrickvonplaten** [pytorch/pytorch#66086]
## 🐛 Bug

The function `torch.cosine_similarity` allowed tensor input of differing shapes in torch 1.9.1. In the nightly version for 1.10, this does not seem to be allowed anymore. This breaks backwards compatibility - is this intended here?

## To Reproduce

1.) Install `torch 1.9.1`:

```
pip3 install torch==1.9.1+cpu torchvision==0.10.1+cpu torchaudio==0.9.1 -f https://download.pytorch.org/whl/torch_stable.html
```

Run:

```python
import torch
torch.cosine_similarity(torch.tensor((10, 2, 3, 4)), torch.tensor((2, 3, 4)), dim=-1)
```

=> works fine

2.) Install `torch-nightly` for `torch 1.10`:

```
pip3 install --pre torch torchvision torchaudio -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
```

Run the same function:

```python
import torch
torch.cosine_similarity(torch.tensor((10, 2, 3, 4)), torch.tensor((2, 3, 4)), dim=-1)
```

=> This function now returns an error:

```
cosine_similarity requires both inputs to have the same size at dimension -1but x1 has 4 and x2 has 3
```

## Expected behavior

IMO backwards compibility should be kept here.

## Environment

 - PyTorch Version (e.g., 1.0): 1.9.1 
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): `pip`
 - Build command you used (if compiling from source): see above
 - Python version: Python 3.9.1
 - CUDA/cuDNN version: CPU
 - GPU models and configuration: /
 - Any other relevant information:

**Comment by malfet** [comment]
Seem to be already addressed by https://github.com/pytorch/pytorch/commit/e6a3154519721eec323b227ea8dfbd05409cfe37

**Comment by malfet** [comment]
https://github.com/pytorch/pytorch/pull/62912 seems related

**Comment by ngimel** [comment]
Thanks for the report, this is not intended. #58559 was a too aggressive check for a real bug in cosine similarity, #62912 relaxed it a bit, but it's still not working for valid inputs like the one in this issue. We'll fix it.

**Comment by ngimel** [comment]
For tensors with the different dimensions, behavior is well defined only for negative `dim` argument, e.g. for the example above `dim=3` would error out and `dim=2` would produce silently wrong results. So I suggest that we allow tensors with different dimensions only when `dim` is negative - does this sound reasonable? Or do we want positive `dim` to indicate dimension of a higher-dimensional tensor, where it's `x1` or `x2`?
